File: gofix/gofix_services/report/gofix_ticket_status_by_location/gofix_ticket_status_by_location.py
# Copyright (c) 2026, GoStack and contributors
# For license information, please see license.txt

"""GoFix Ticket Status by Location.

One row per Service Request with its full lifecycle position — SR status,
device transfer/location, technician, QC verdict, billing — grouped by the
store the ticket was raised at. Answers "what is the complete status of every
ticket, location wise" in a single screen.
"""

import frappe
from ch_erp15.ch_erp15.report_scope import scope_where_clause
from frappe import _
from frappe.utils import date_diff, flt, today


def execute(filters=None):
	filters = frappe._dict(filters or {})
	columns = get_columns()
	data = get_data(filters)
	summary = get_summary(data)
	chart = get_chart(data)
	return columns, data, None, chart, summary


def get_columns():
	return [
		{"label": _("Location"), "fieldname": "location", "fieldtype": "Link", "options": "Warehouse", "width": 200},
		{"label": _("Ticket"), "fieldname": "name", "fieldtype": "Link", "options": "Service Request", "width": 150},
		{"label": _("Date"), "fieldname": "service_date", "fieldtype": "Date", "width": 95},
		{"label": _("Customer"), "fieldname": "customer_name", "fieldtype": "Data", "width": 140},
		{"label": _("Device"), "fieldname": "device", "fieldtype": "Data", "width": 170},
		{"label": _("Issue"), "fieldname": "issue_category", "fieldtype": "Link", "options": "Issue Category", "width": 120},
		{"label": _("Priority"), "fieldname": "priority", "fieldtype": "Data", "width": 75},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 105},
		{"label": _("Device At"), "fieldname": "device_at", "fieldtype": "Data", "width": 170},
		{"label": _("Transfer"), "fieldname": "transfer_status", "fieldtype": "Data", "width": 130},
		{"label": _("Technician"), "fieldname": "technician", "fieldtype": "Data", "width": 130},
		{"label": _("QC"), "fieldname": "qc_status", "fieldtype": "Data", "width": 70},
		# Where a finished device is, and who is moving it. Without these the
		# report could say a repair was done but not whether anyone was coming
		# for it -- which is the question the shop floor actually asks.
		{"label": _("Goes Back By"), "fieldname": "return_method", "fieldtype": "Link",
		 "options": "Device Logistics Method", "width": 130},
		{"label": _("Carrier"), "fieldname": "return_partner", "fieldtype": "Link",
		 "options": "Courier Partner", "width": 110},
		{"label": _("Tracking"), "fieldname": "return_tracking_number",
		 "fieldtype": "Data", "width": 130},
		{"label": _("Dispatched"), "fieldname": "return_dispatched_date",
		 "fieldtype": "Date", "width": 100},
		{"label": _("Estimate"), "fieldname": "estimated_cost", "fieldtype": "Currency", "width": 100},
		{"label": _("Invoice"), "fieldname": "service_invoice", "fieldtype": "Link", "options": "Sales Invoice", "width": 140},
		{"label": _("Billed"), "fieldname": "billed_amount", "fieldtype": "Currency", "width": 100},
		{"label": _("Completed"), "fieldname": "actual_completion_date", "fieldtype": "Date", "width": 95},
		{"label": _("Days Open"), "fieldname": "days_open", "fieldtype": "Int", "width": 85},
		{"label": _("Timeline"), "fieldname": "timeline_btn", "fieldtype": "Data", "width": 90},
	]


def get_data(filters):
	conditions = ["sr.docstatus < 2"]
	values = {}

	if filters.get("company"):
		conditions.append("sr.company = %(company)s")
		values["company"] = filters.company
	if filters.get("location"):
		conditions.append("sr.source_warehouse = %(location)s")
		values["location"] = filters.location
	if filters.get("from_date"):
		conditions.append("sr.service_date >= %(from_date)s")
		values["from_date"] = filters.from_date
	if filters.get("to_date"):
		conditions.append("sr.service_date <= %(to_date)s")
		values["to_date"] = filters.to_date
	if filters.get("status"):
		conditions.append("sr.decision = %(status)s")
		values["status"] = filters.status
	if not filters.get("include_closed"):
		conditions.append("sr.decision NOT IN ('Cancelled', 'Rejected')")

	# Row-level scope: tickets raised at a store within the user's scope.
	scope = scope_where_clause(warehouse_field="sr.source_warehouse")
	if scope:
		conditions.append(scope)

	# The geography the user asked for, on top of the scope they are entitled
	# to. Asked and allowed are AND-ed: choosing a zone you cannot see returns
	# nothing, never everything.
	from gofix.report_filters import geo_conditions

	_geo = geo_conditions(filters, company_field='sr.company', warehouse_field='sr.source_warehouse')
	if _geo:
		conditions.append(_geo[len(" AND "):])

	rows = frappe.db.sql(
		f"""
		SELECT
			sr.name, sr.source_warehouse AS location, sr.service_date,
			sr.customer_name, sr.customer,
			COALESCE(NULLIF(sr.device_item_name, ''), sr.device_item) AS device,
			sr.issue_category, sr.priority, sr.decision AS status, sr.decision,
			sr.transfer_status, sr.current_location, sr.transferred_to_store,
			sr.estimated_cost, sr.service_invoice, sr.service_order,
			sr.return_method, sr.return_partner, sr.return_tracking_number,
			sr.return_dispatched_date, sr.intake_method,
			sr.actual_completion_date,
			-- The request is the QC document; the order is the legacy fallback.
			COALESCE(NULLIF(sr.qc_status, ''), so.qc_status) AS qc_status,
			si.grand_total AS billed_amount, si.status AS invoice_status,
			(
				SELECT emp.employee_name
				FROM `tabJob Assignment` ja
				LEFT JOIN `tabEmployee` emp ON emp.name = ja.service_engineer
				WHERE ja.service_request = sr.name AND ja.docstatus < 2
				ORDER BY ja.creation DESC LIMIT 1
			) AS technician
		FROM `tabService Request` sr
		LEFT JOIN `tabSales Order` so ON so.name = sr.service_order
		LEFT JOIN `tabSales Invoice` si ON si.name = sr.service_invoice
		WHERE {" AND ".join(conditions)}
		ORDER BY sr.source_warehouse, sr.service_date DESC, sr.name
		""",
		values,
		as_dict=True,
	)

	for r in rows:
		# Where the device physically is right now
		if r.transfer_status == "In Transit":
			r.device_at = f"In transit → {r.transferred_to_store or ''}"
		elif r.transfer_status == "Return In Transit":
			r.device_at = f"In transit → {r.location}"
		else:
			r.device_at = r.current_location or r.location
		if r.invoice_status:
			r.status = f"{r.status} ({r.invoice_status})" if r.service_invoice else r.status
		end = r.actual_completion_date or today()
		r.days_open = max(date_diff(end, r.service_date), 0) if r.service_date else None
		r.timeline_btn = "View"

	return rows


def get_summary(data):
	total = len(data)
	completed = sum(1 for r in data if (r.status or "").startswith(("Completed", "Invoiced", "Delivered")))
	in_progress = total - completed
	billed = sum(flt(r.billed_amount) for r in data)
	locations = len({r.location for r in data})
	return [
		{"label": _("Locations"), "value": locations, "datatype": "Int"},
		{"label": _("Tickets"), "value": total, "datatype": "Int"},
		{"label": _("In Progress"), "value": in_progress, "datatype": "Int", "indicator": "Orange"},
		{"label": _("Completed"), "value": completed, "datatype": "Int", "indicator": "Green"},
		{"label": _("Billed Amount"), "value": billed, "datatype": "Currency"},
	]


def get_chart(data):
	buckets = {}
	for r in data:
		loc = r.location or "—"
		b = buckets.setdefault(loc, {"open": 0, "done": 0})
		if (r.status or "").startswith(("Completed", "Invoiced", "Delivered")):
			b["done"] += 1
		else:
			b["open"] += 1
	labels = sorted(buckets)
	return {
		"data": {
			"labels": labels,
			"datasets": [
				{"name": _("In Progress"), "values": [buckets[l]["open"] for l in labels]},
				{"name": _("Completed"), "values": [buckets[l]["done"] for l in labels]},
			],
		},
		"type": "bar",
		"barOptions": {"stacked": 1},
	}

File: gofix/gofix_services/report/gofix_ticket_status_by_location/gofix_ticket_status_by_location.js
// Copyright (c) 2026, GoStack and contributors
// For license information, please see license.txt

frappe.query_reports["GoFix Ticket Status by Location"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "location",
			label: __("Location (Store)"),
			fieldtype: "Link",
			options: "Warehouse",
			get_query: () => ({
				filters: { is_group: 0, disabled: 0 },
			}),
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
		},
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options: [
				"", "Draft", "Open", "In Service", "Waiting for Parts",
				"Ready for Delivery", "Completed", "Invoiced", "Delivered",
				"Cancelled", "Rejected",
			].join("\n"),
		},
		{
			fieldname: "include_closed",
			label: __("Include Cancelled / Rejected"),
			fieldtype: "Check",
			default: 0,
		},
	],

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "status" && data) {
			const s = String(data.status || "");
			let color = "orange";
			if (s.startsWith("Completed") || s.startsWith("Invoiced") || s.startsWith("Delivered")) color = "green";
			else if (s.startsWith("Cancelled") || s.startsWith("Rejected")) color = "red";
			value = `<span class="indicator-pill ${color}">${value}</span>`;
		}
		if (column.fieldname === "qc_status" && data && data.qc_status) {
			const color = data.qc_status === "Pass" ? "green" : data.qc_status === "Fail" ? "red" : "orange";
			value = `<span class="indicator-pill ${color}">${value}</span>`;
		}
		if (column.fieldname === "days_open" && data && data.days_open > 7) {
			value = `<span style="color:var(--red-600);font-weight:600">${value}</span>`;
		}
		if (column.fieldname === "timeline_btn" && data && data.name) {
			value = `<a href="#" onclick="gofix_ticket_timeline('${frappe.utils.escape_html(data.name)}'); return false;"
				style="font-weight:600">📈 ${__("View")}</a>`;
		}
		return value;
	},
};

// ── Per-ticket timeline dialog ──────────────────────────────────────────────

const GOFIX_TL_COLORS = [
	"#2563eb", "#16a34a", "#d97706", "#dc2626", "#7c3aed",
	"#0891b2", "#be185d", "#65a30d", "#b45309", "#4f46e5",
];

function gofix_tl_parse(dt) {
	return new Date(String(dt).replace(" ", "T"));
}

function gofix_tl_duration(ms) {
	if (ms < 0) ms = 0;
	const mins = Math.round(ms / 60000);
	if (mins < 1) return __("moments");
	if (mins < 60) return `${mins}m`;
	const hrs = Math.floor(mins / 60);
	if (hrs < 24) return `${hrs}h ${mins % 60}m`;
	const days = Math.floor(hrs / 24);
	return `${days}d ${hrs % 24}h`;
}

window.gofix_ticket_timeline = async function (sr_name) {
	let events;
	try {
		events = await frappe.xcall(
			"gofix.gofix_services.page.gofix_ops_hub.gofix_ops_hub.get_repair_history",
			{ sr_name }
		);
	} catch (e) {
		frappe.msgprint({ title: __("Timeline"), message: e.message || String(e), indicator: "red" });
		return;
	}
	if (!events || !events.length) {
		frappe.msgprint(__("No timeline events recorded for {0}.", [sr_name]));
		return;
	}

	const esc = frappe.utils.escape_html;
	const times = events.map((e) => gofix_tl_parse(e.at));
	const t0 = times[0];
	const tN = times[times.length - 1];
	const total = Math.max(tN - t0, 1);

	// ── Phase-duration bar: one segment per gap between consecutive events ──
	let segments = "";
	for (let i = 1; i < events.length; i++) {
		const gap = times[i] - times[i - 1];
		if (gap <= 0) continue;
		const pct = Math.max((gap / total) * 100, 0.8);
		const color = GOFIX_TL_COLORS[(i - 1) % GOFIX_TL_COLORS.length];
		const tip = `${esc(events[i - 1].title)} → ${esc(events[i].title)}: ${gofix_tl_duration(gap)}`;
		segments += `<div title="${tip}" style="width:${pct}%;background:${color};height:100%;"></div>`;
	}
	const bar = `
		<div style="margin-bottom:4px;font-size:12px;color:var(--text-muted);">
			${__("Total turnaround")}: <b style="color:var(--text-color)">${gofix_tl_duration(total)}</b>
			&nbsp;·&nbsp; ${frappe.datetime.str_to_user(events[0].at)} → ${frappe.datetime.str_to_user(events[events.length - 1].at)}
			&nbsp;·&nbsp; ${events.length} ${__("events")}
		</div>
		<div style="display:flex;height:18px;border-radius:9px;overflow:hidden;border:1px solid var(--border-color);margin-bottom:16px;">
			${segments || `<div style="width:100%;background:var(--control-bg)"></div>`}
		</div>
		<div style="font-size:11px;color:var(--text-muted);margin:-10px 0 14px;">
			${__("Hover a segment to see the phase and how long it took.")}
		</div>`;

	// ── Vertical event timeline ─────────────────────────────────────────────
	let rows = "";
	for (let i = 0; i < events.length; i++) {
		const ev = events[i];
		const color = i === 0 ? "var(--gray-500)" : GOFIX_TL_COLORS[(i - 1) % GOFIX_TL_COLORS.length];
		const gap = i > 0 ? `<span style="color:var(--text-muted);font-size:11px;">&nbsp;(+${gofix_tl_duration(times[i] - times[i - 1])})</span>` : "";
		const ref = ev.ref_name && ev.ref_doctype
			? `&nbsp;<a href="/app/${frappe.router.slug(ev.ref_doctype)}/${encodeURIComponent(ev.ref_name)}"
					style="font-size:11px;">${esc(ev.ref_name)}</a>`
			: "";
		const detail = ev.detail ? `<div style="color:var(--text-muted);font-size:12px;margin-top:1px;">${esc(ev.detail)}</div>` : "";
		rows += `
			<div style="display:flex;gap:12px;">
				<div style="display:flex;flex-direction:column;align-items:center;">
					<div style="width:12px;height:12px;border-radius:50%;background:${color};margin-top:3px;flex-shrink:0;"></div>
					${i < events.length - 1 ? `<div style="width:2px;flex:1;background:var(--border-color);min-height:14px;"></div>` : ""}
				</div>
				<div style="padding-bottom:14px;min-width:0;">
					<div style="font-size:11px;color:var(--text-muted);white-space:nowrap;">
						${frappe.datetime.str_to_user(ev.at)}${gap}
					</div>
					<div style="font-weight:600;font-size:13px;">${esc(ev.title)}${ref}</div>
					${detail}
				</div>
			</div>`;
	}

	const d = new frappe.ui.Dialog({
		title: __("Repair Timeline — {0}", [sr_name]),
		size: "large",
		fields: [{ fieldtype: "HTML", fieldname: "tl" }],
		primary_action_label: __("Open Ticket"),
		primary_action() {
			frappe.set_route("Form", "Service Request", sr_name);
			d.hide();
		},
	});
	d.fields_dict.tl.$wrapper.html(`<div style="max-height:65vh;overflow-y:auto;padding-right:6px;">${bar}${rows}</div>`);
	d.show();
};
